src/py/mat3ra/made/tools/build/passivation/coordination.py

```diff
@@ -0,0 +1,195 @@
+
+    angle_tolerance: float = Field(0.1, description="Tolerance for comparing angles between bond vectors.")
+    max_bonds_to_passivate: int = Field(
+        2, description="Maximum number of bonds to passivate for each undercoordinated atom."
+    )
+
+    @staticmethod
+    def are_bonds_templates_similar(template1: np.ndarray, template2: np.ndarray, tolerance: float = 0.1) -> bool:
+        """
+        Check if two bond templates are similar.
+
+        Args:
+            template1 (np.ndarray): First template of bond vectors.
+            template2 (np.ndarray): Second template of bond vectors.
+            tolerance (float): Angle tolerance for comparison.
+
+        Returns:
+            bool: True if the templates are similar, False otherwise.
+        """
+        if len(template1) != len(template2):
+            return False
+
+        dot_matrix = np.dot(template1, template2.T)
+        norms1 = np.linalg.norm(template1, axis=1)
+        norms2 = np.linalg.norm(template2, axis=1)
+        cosine_matrix = dot_matrix / np.outer(norms1, norms2)
+        angles_matrix = np.arccos(np.clip(cosine_matrix, -1.0, 1.0))
+
+        unmatched = list(range(len(template2)))
+        for angle_row in angles_matrix:
+            matches = np.where(angle_row < tolerance)[0]
+            if len(matches) == 0:
+                return False
+            unmatched.remove(matches.tolist()[0])
+
+        return True
+
+    def find_template_vectors(
+        self, atom_vectors: List[List[np.ndarray]], atom_elements: List[str]
+    ) -> Dict[str, List[np.ndarray]]:
+        """
+        Find unique bond templates for each element type.
+
+        Args:
+            atom_vectors (List[List[np.ndarray]]): List of bond vectors for each atom.
+            atom_elements (List[str]): List of chemical elements for each atom.
+
+        Returns:
+            Dict[str, List[np.ndarray]]: Dictionary mapping element types to unique bond templates.
+        """
+        element_templates = {}
+
+        for element in set(atom_elements):
+            element_indices = [i for i, e in enumerate(atom_elements) if e == element]
+            element_vector_lists = [np.array(atom_vectors[i]) for i in element_indices]
+
+            if not element_vector_lists:
+                continue
+
+            max_coord = max(len(vectors) for vectors in element_vector_lists)
+            max_coord_vectors = [v for v in element_vector_lists if len(v) == max_coord]
+
+            unique_templates: List[np.ndarray] = []
+            for template in max_coord_vectors:
+                if not any(
+                    self.are_bonds_templates_similar(template, existing, self.angle_tolerance)
+                    for existing in unique_templates
+                ):
+                    unique_templates.append(template)
+
+            element_templates[element] = unique_templates
+
+        return element_templates
+
+    def reconstruct_missing_bonds(
```

Comment:
isolate reconstruct_missing_bonds_for_atom and have `reconstruct_missing_bonds` for the whole material

Reply:
Ok. And this is where we'll pass bond templates distinguished by element